app/src/main/java/com/github/swent/echo/data/supabase/SupabaseDataSource.kt

```diff
@@ -63,112 +70,271 @@ class SupabaseDataSource(private val supabase: SupabaseClient) : RemoteDataSourc
                 }
                 .decodeSingle<AssociationSupabase>()
                 .toAssociation()
-        } catch (e: BadRequestRestException) {
-            null
+        } catch (e: Exception) {
+            when (e) {
+                is BadRequestRestException -> null
+                is HttpRequestException -> {
+                    if (maxRetriesCount == 0u) throw e
+                    delay(RETRY_DELAY_MILLI)
+                    return getAssociation(associationId, maxRetriesCount - 1u)
+                }
+                else -> throw e
```

Comment:
You could write a function which wraps the supabase call and reuse it everywhere. It could look like this:
```kotlin
/**
 * Wrapper for Supabase requests that retries the request if it fails due to a
 * [HttpRequestException]. If the request fails after [maxRetriesCount] retries, it throws the
 * exception. In case of an [NoSuchElementException], it returns null.
 *
 * @param maxRetriesCount: the number of times the request will be retried
 * @param request: the request to be executed
 */
private suspend fun <T> supabaseRequestWrapper(
    maxRetriesCount: UInt,
    request: suspend () -> T,
): T? {
    return try {
        request()
    } catch (e: NoSuchElementException) {
        null
    } catch (e: HttpRequestException) {
        if (maxRetriesCount == 0u) throw e
        delay(RETRY_DELAY_MILLI)
        return supabaseRequestWrapper(maxRetriesCount - 1u, request)
    }
}
```

And you could use it like this:
```kotlin
override suspend fun getAssociation(
    associationId: String,
    maxRetriesCount: UInt
): Association? = supabaseRequestWrapper(maxRetriesCount) {
    supabase
        .from("associations")
        .select(Columns.raw(QUERY_ASSOCIATION)) {
            filter { eq("association_id", associationId) }
        }
        .decodeSingle<AssociationSupabase>()
        .toAssociation()
}
```

Note that you would need to also write a wrapper fur function which return a list.

Reply:
Thanks! Yes absolutely. I've been thinking about generalizing these. Actually there are quite a few things which are similar for each type of methods (single getters, getAlls, deletes, etc). However there are still some subtle differences for some of them and I'm unsure up to which point generalization makes sense. Even for the wrapper you're suggesting, there are actually some functions for which I check for different Exceptions.
I'll look into this and see if I can build something reasonable without breaking the tests :D

Reply:
Okay, so I heavily reworked this. Based on your suggestion I did a wrapper function for the retry and exception handling in the SupabaseDataSource. In a similar way, i reworked the code from the RepositoryImpl to remove redundancies as much as possible. (The changes are in the last two commits)